pkg/controllers/customroute/custom_route_controller.go

```diff
@@ -0,0 +1,509 @@
+	if componentRoute := common.GetComponentRouteSpec(ingressConfig, OAuthComponentRouteNamespace, OAuthComponentRouteName); componentRoute != nil {
+		var errors []error
+		// Check if the provided secret is valid
+		secretName = componentRoute.ServingCertKeyPairSecret.Name
+		if err := c.validateCustomTLSSecret(secretName); err != nil {
+			errors = append(errors, err)
+		}
+
+		// Check if the provided hostname is valid
+		hostname := string(componentRoute.Hostname)
+		if err := validateCustomHostname(hostname); err != nil {
+			errors = append(errors, err)
+		}
+
+		if errors != nil {
+			return nil, "", errors
+		}
+
+		route.Spec.Host = hostname
+	}
+
+	return route, secretName, nil
+}
+
+func (c *customRouteController) validateCustomTLSSecret(secretName string) error {
+	if secretName != "" {
+		secret, err := c.secretLister.Secrets("openshift-config").Get(secretName)
+		if err != nil {
+			return err
+		}
+
+		var errors []error
+		privateKeyData, ok := secret.Data[corev1.TLSPrivateKeyKey]
+		if !ok {
+			errors = append(errors, fmt.Errorf("custom route secret must include key %s", corev1.TLSPrivateKeyKey))
+		} else {
+			errors = append(errors, datasync.ValidatePrivateKey(privateKeyData)...)
+		}
+
+		certData, ok := secret.Data[corev1.TLSCertKey]
+		if !ok {
+			errors = append(errors, fmt.Errorf("custom route secret must include key %s", corev1.TLSCertKey))
+		} else {
+			errors = append(errors, datasync.ValidateServerCert(certData)...)
+		}
+
+		if len(errors) != 0 {
+			return fmt.Errorf("error validating secret %s/%s: %v", "openshift-config", secretName, errors)
+		}
+	}
+	return nil
+}
+
+func validateCustomHostname(hostname string) error {
+	re, err := regexp.Compile(hostnameRegex)
+	if err != nil {
+		return err
+	}
+
+	if !re.MatchString(hostname) {
+		return fmt.Errorf("provided hostname (%s) is invalid, must match %s", hostname, hostnameRegex)
+	}
+
+	return nil
+}
+
+func (c *customRouteController) applyRoute(ctx context.Context, expectedRoute *routev1.Route) error {
+	route, err := c.routeClient.Get(ctx, "oauth-openshift", metav1.GetOptions{})
+	if errors.IsNotFound(err) {
+		route, err = c.routeClient.Create(ctx, expectedRoute, metav1.CreateOptions{})
+	}
+	if err != nil {
+		return err
+	}
+
+	// assume it is unsafe to mutate route in case we go to a shared informer in the future
+	existingCopy := route.DeepCopy()
+	modified := resourcemerge.BoolPtr(false)
+	resourcemerge.EnsureObjectMeta(modified, &existingCopy.ObjectMeta, expectedRoute.ObjectMeta)
+
+	// this guarantees that route.Spec.Host is set to the current canonical host
+	if *modified || !equality.Semantic.DeepEqual(existingCopy.Spec, expectedRoute.Spec) {
+		// be careful not to print route.spec as it many contain secrets
+		existingCopy.Spec = expectedRoute.Spec
+		route, err = c.routeClient.Update(ctx, existingCopy, metav1.UpdateOptions{})
+		if err != nil {
+			return err
+		}
+	}
+
+	_, _, err = routeapihelpers.IngressURI(route, expectedRoute.Spec.Host)
```

Comment:
I might have asked this before and this might have been answered, but - why is this here?

Reply:
This already existed when I pulled it into the package, I'll remove it.